From d11ceabaecdd379d8e05177eae2e6453226a7e57 Mon Sep 17 00:00:00 2001
From: Shammah Chancellor <shammah.chancellor@gmail.com>
Date: Sat, 19 Jun 2021 11:01:14 -0700
Subject: [PATCH 122/196] Partially revert "[eCash] add the XEC base unit and
 test it by default"

This reverts commit 6668fc39897aa18fae99ba09a0dac8da02d66297, and adds
in the necessary parameters to ensure that the unit tests -- which were
all updated to use the `ecash` prefix -- still pass.
---
 src/amount.cpp                              |   1 -
 src/chainparams.cpp                         |  11 +-
 src/currencyunit.cpp                        |   4 +
 src/qt/test/uritests.cpp                    |   7 +-
 src/test/amount_tests.cpp                   |   2 +-
 src/test/data/script_tests.json             |  10 +-
 src/test/rpc_tests.cpp                      | 109 ++++++++++----------
 src/test/util/setup_common.cpp              |   2 +-
 src/test/util_tests.cpp                     | 106 ++++++++++---------
 test/functional/test_framework/test_node.py |   1 +
 10 files changed, 135 insertions(+), 118 deletions(-)

diff --git a/src/amount.cpp b/src/amount.cpp
index 91ab8c1f4..108492b9d 100644
--- a/src/amount.cpp
+++ b/src/amount.cpp
@@ -10,7 +10,6 @@
 #include <network.h>
 #include <univalue.h>
 #include <util/system.h>
-
 #include <tinyformat.h>
 
 static const Currency BCHA{COIN, SATOSHI, 8, "BCHA"};
diff --git a/src/chainparams.cpp b/src/chainparams.cpp
index 2acc1e69e..1f01ae4af 100644
--- a/src/chainparams.cpp
+++ b/src/chainparams.cpp
@@ -17,6 +17,10 @@
 
 #include <cassert>
 
+static bool UseECashPrefix() {
+    return gArgs.GetBoolArg("-ecash", DEFAULT_ECASH) || gArgs.GetBoolArg("-useecashprefix", DEFAULT_ECASH);
+}
+
 static CBlock CreateGenesisBlock(const char *pszTimestamp,
                                  const CScript &genesisOutputScript,
                                  uint32_t nTime, uint32_t nNonce,
@@ -204,8 +208,9 @@ public:
         base58Prefixes[SECRET_KEY] = std::vector<uint8_t>(1, 128);
         base58Prefixes[EXT_PUBLIC_KEY] = {0x04, 0x88, 0xB2, 0x1E};
         base58Prefixes[EXT_SECRET_KEY] = {0x04, 0x88, 0xAD, 0xE4};
+
         cashaddrPrefix =
-            gArgs.GetBoolArg("-ecash", DEFAULT_ECASH) ? "ecash" : "bitcoincash";
+            UseECashPrefix() ? "ecash" : "bitcoincash";
 
         vFixedSeeds = std::vector<SeedSpec6>(
             pnSeed6_main, pnSeed6_main + ARRAYLEN(pnSeed6_main));
@@ -353,7 +358,7 @@ public:
         base58Prefixes[EXT_PUBLIC_KEY] = {0x04, 0x35, 0x87, 0xCF};
         base58Prefixes[EXT_SECRET_KEY] = {0x04, 0x35, 0x83, 0x94};
         cashaddrPrefix =
-            gArgs.GetBoolArg("-ecash", DEFAULT_ECASH) ? "ectest" : "bchtest";
+            UseECashPrefix() ? "ectest" : "bchtest";
 
         vFixedSeeds = std::vector<SeedSpec6>(
             pnSeed6_test, pnSeed6_test + ARRAYLEN(pnSeed6_test));
@@ -486,7 +491,7 @@ public:
         base58Prefixes[EXT_PUBLIC_KEY] = {0x04, 0x35, 0x87, 0xCF};
         base58Prefixes[EXT_SECRET_KEY] = {0x04, 0x35, 0x83, 0x94};
         cashaddrPrefix =
-            gArgs.GetBoolArg("-ecash", DEFAULT_ECASH) ? "ecregtest" : "bchreg";
+            UseECashPrefix() ? "ecregtest" : "bchreg";
     }
 };
 
diff --git a/src/currencyunit.cpp b/src/currencyunit.cpp
index 7c944c9c8..bd122f73c 100644
--- a/src/currencyunit.cpp
+++ b/src/currencyunit.cpp
@@ -12,4 +12,8 @@ void SetupCurrencyUnitOptions(ArgsManager &argsman) {
                    strprintf("Use the eCash prefixes and units (default: %s)",
                              DEFAULT_ECASH ? "true" : "false"),
                    ArgsManager::ALLOW_BOOL, OptionsCategory::OPTIONS);
+    argsman.AddArg("-useecashprefix",
+               strprintf("Use the eCash prefixes (default: %s)",
+                         DEFAULT_ECASH ? "true" : "false"),
+               ArgsManager::ALLOW_BOOL, OptionsCategory::OPTIONS);
 }
diff --git a/src/qt/test/uritests.cpp b/src/qt/test/uritests.cpp
index 63473d6f0..f74792f97 100644
--- a/src/qt/test/uritests.cpp
+++ b/src/qt/test/uritests.cpp
@@ -44,7 +44,7 @@ void URITests::uriTestsCashAddr() {
     QVERIFY(rv.address ==
             QString("ecash:qpm2qsznhks23z7629mms6s4cwef74vcwva87rkuu2"));
     QVERIFY(rv.label == QString());
-    QVERIFY(rv.amount == SATOSHI);
+    QVERIFY(rv.amount == 1000000 * SATOSHI);
 
     uri.setUrl(QString(
         "ecash:qpm2qsznhks23z7629mms6s4cwef74vcwva87rkuu2?amount=10.01"));
@@ -52,7 +52,7 @@ void URITests::uriTestsCashAddr() {
     QVERIFY(rv.address ==
             QString("ecash:qpm2qsznhks23z7629mms6s4cwef74vcwva87rkuu2"));
     QVERIFY(rv.label == QString());
-    QVERIFY(rv.amount == 1001 * SATOSHI);
+    QVERIFY(rv.amount == int64_t(1001000000) * SATOSHI);
 
     uri.setUrl(
         QString("ecash:qpm2qsznhks23z7629mms6s4cwef74vcwva87rkuu2?amount=100&"
@@ -60,7 +60,8 @@ void URITests::uriTestsCashAddr() {
     QVERIFY(GUIUtil::parseBitcoinURI(scheme, uri, &rv));
     QVERIFY(rv.address ==
             QString("ecash:qpm2qsznhks23z7629mms6s4cwef74vcwva87rkuu2"));
-    QVERIFY(rv.amount == 10000 * SATOSHI);
+                std::cout << rv.amount;
+    QVERIFY(rv.amount == int64_t(10000000000) * SATOSHI);
     QVERIFY(rv.label == QString("Wikipedia Example"));
 
     uri.setUrl(
diff --git a/src/test/amount_tests.cpp b/src/test/amount_tests.cpp
index c2f4d2331..cb40d9553 100644
--- a/src/test/amount_tests.cpp
+++ b/src/test/amount_tests.cpp
@@ -132,7 +132,7 @@ BOOST_AUTO_TEST_CASE(ToStringTest) {
     CFeeRate feeRate;
     feeRate = CFeeRate(1 * SATOSHI);
     BOOST_CHECK_EQUAL(feeRate.ToString(),
-                      "0.01 " + Currency::get().ticker + "/kB");
+                      "0.00000001 " + Currency::get().ticker + "/kB");
 }
 
 BOOST_AUTO_TEST_SUITE_END()
diff --git a/src/test/data/script_tests.json b/src/test/data/script_tests.json
index 6b8feb5d9..240369854 100644
--- a/src/test/data/script_tests.json
+++ b/src/test/data/script_tests.json
@@ -2486,7 +2486,7 @@
 ],
 [
     [
-        123450000000.00
+        123450.00000000
     ],
     "0x47 0x304402206e3afa6dd4d1db87538fa48a0ef3f824d7ec554103fe5fb3527254d78bd79617022066097e981df0d1ce3c07224a73585ca1f5e8ebdf79639b2cc73c20bde066075141",
     "0x41 0x0479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8 CHECKSIG",
@@ -2496,7 +2496,7 @@
 ],
 [
     [
-        123450000000.00
+        123450.00000000
     ],
     "0x47 0x3044022054e19540c81ccb214b62acda158faf01440d98adab2e13ed43a4d35cc551bfff02205b05fc09882819b3fdbe56adfef63243d7faa5c4837db0c72ad9356eae9a5b8c41",
     "0x41 0x0479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8 CHECKSIG",
@@ -2506,7 +2506,7 @@
 ],
 [
     [
-        123450000000.00
+        123450.00000000
     ],
     "0x47 0x304402206e3afa6dd4d1db87538fa48a0ef3f824d7ec554103fe5fb3527254d78bd79617022066097e981df0d1ce3c07224a73585ca1f5e8ebdf79639b2cc73c20bde066075141",
     "0x41 0x0479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8 CHECKSIG",
@@ -2516,7 +2516,7 @@
 ],
 [
     [
-        123450000000.00
+        123450.00000000
     ],
     "0x47 0x304402205ce75e7249823fe2cc78f861579189f8e10585008aa057d4e6d755bfca1c8c6802200f63547feabc3f198f18742e84d57d6ca4bdd477d10a121554647cba36742f2b41",
     "0x41 0x0479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8 CHECKSIG",
@@ -2526,7 +2526,7 @@
 ],
 [
     [
-        123450000000.00
+        123450.00000000
     ],
     "0x47 0x304402206e3afa6dd4d1db87538fa48a0ef3f824d7ec554103fe5fb3527254d78bd79617022066097e981df0d1ce3c07224a73585ca1f5e8ebdf79639b2cc73c20bde066075141",
     "0x41 0x0479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8 CHECKSIG",
diff --git a/src/test/rpc_tests.cpp b/src/test/rpc_tests.cpp
index d8eb6084e..143b5a9e7 100644
--- a/src/test/rpc_tests.cpp
+++ b/src/test/rpc_tests.cpp
@@ -134,7 +134,7 @@ BOOST_AUTO_TEST_CASE(rpc_rawsign) {
                           "8f50977c8493f3\","
                           "\"vout\":1,\"scriptPubKey\":"
                           "\"a914b10c9df5f7edf436c697f02f1efdba4cf399615187\","
-                          "\"amount\":3141590.00,"
+                          "\"amount\":3.14159,"
                           "\"redeemScript\":"
                           "\"512103debedc17b3df2badbcdd86d5feb4562b86fe182e5998"
                           "abd8bcd4f122c6155b1b21027e940bb73ab8732bfdf7f9216ece"
@@ -237,40 +237,40 @@ BOOST_AUTO_TEST_CASE(rpc_createraw_op_return) {
 }
 
 BOOST_AUTO_TEST_CASE(rpc_format_monetary_values) {
-    BOOST_CHECK(UniValue(Amount::zero()).write() == "0.00");
-    BOOST_CHECK(UniValue(SATOSHI).write() == "0.01");
-    BOOST_CHECK(UniValue(17622195 * SATOSHI).write() == "176221.95");
-    BOOST_CHECK(UniValue(50000000 * SATOSHI).write() == "500000.00");
-    BOOST_CHECK(UniValue(89898989 * SATOSHI).write() == "898989.89");
-    BOOST_CHECK(UniValue(100000000 * SATOSHI).write() == "1000000.00");
+    BOOST_CHECK(UniValue(Amount::zero()).write() == "0.00000000");
+    BOOST_CHECK(UniValue(SATOSHI).write() == "0.00000001");
+    BOOST_CHECK(UniValue(17622195 * SATOSHI).write() == "0.17622195");
+    BOOST_CHECK(UniValue(50000000 * SATOSHI).write() == "0.50000000");
+    BOOST_CHECK(UniValue(89898989 * SATOSHI).write() == "0.89898989");
+    BOOST_CHECK(UniValue(100000000 * SATOSHI).write() == "1.00000000");
     BOOST_CHECK(UniValue(int64_t(2099999999999990) * SATOSHI).write() ==
-                "20999999999999.90");
+                "20999999.99999990");
     BOOST_CHECK(UniValue(int64_t(2099999999999999) * SATOSHI).write() ==
-                "20999999999999.99");
+                "20999999.99999999");
 
-    BOOST_CHECK_EQUAL(UniValue(Amount::zero()).write(), "0.00");
+    BOOST_CHECK_EQUAL(UniValue(Amount::zero()).write(), "0.00000000");
     BOOST_CHECK_EQUAL(UniValue(123456789 * (COIN / 10000)).write(),
-                      "12345678900.00");
-    BOOST_CHECK_EQUAL(UniValue(-1 * COIN).write(), "-1000000.00");
-    BOOST_CHECK_EQUAL(UniValue(-1 * COIN / 10).write(), "-100000.00");
-
-    BOOST_CHECK_EQUAL(UniValue(100000000 * COIN).write(), "100000000000000.00");
-    BOOST_CHECK_EQUAL(UniValue(10000000 * COIN).write(), "10000000000000.00");
-    BOOST_CHECK_EQUAL(UniValue(1000000 * COIN).write(), "1000000000000.00");
-    BOOST_CHECK_EQUAL(UniValue(100000 * COIN).write(), "100000000000.00");
-    BOOST_CHECK_EQUAL(UniValue(10000 * COIN).write(), "10000000000.00");
-    BOOST_CHECK_EQUAL(UniValue(1000 * COIN).write(), "1000000000.00");
-    BOOST_CHECK_EQUAL(UniValue(100 * COIN).write(), "100000000.00");
-    BOOST_CHECK_EQUAL(UniValue(10 * COIN).write(), "10000000.00");
-    BOOST_CHECK_EQUAL(UniValue(COIN).write(), "1000000.00");
-    BOOST_CHECK_EQUAL(UniValue(COIN / 10).write(), "100000.00");
-    BOOST_CHECK_EQUAL(UniValue(COIN / 100).write(), "10000.00");
-    BOOST_CHECK_EQUAL(UniValue(COIN / 1000).write(), "1000.00");
-    BOOST_CHECK_EQUAL(UniValue(COIN / 10000).write(), "100.00");
-    BOOST_CHECK_EQUAL(UniValue(COIN / 100000).write(), "10.00");
-    BOOST_CHECK_EQUAL(UniValue(COIN / 1000000).write(), "1.00");
-    BOOST_CHECK_EQUAL(UniValue(COIN / 10000000).write(), "0.10");
-    BOOST_CHECK_EQUAL(UniValue(COIN / 100000000).write(), "0.01");
+                      "12345.67890000");
+    BOOST_CHECK_EQUAL(UniValue(-1 * COIN).write(), "-1.00000000");
+    BOOST_CHECK_EQUAL(UniValue(-1 * COIN / 10).write(), "-0.10000000");
+
+    BOOST_CHECK_EQUAL(UniValue(100000000 * COIN).write(), "100000000.00000000");
+    BOOST_CHECK_EQUAL(UniValue(10000000 * COIN).write(), "10000000.00000000");
+    BOOST_CHECK_EQUAL(UniValue(1000000 * COIN).write(), "1000000.00000000");
+    BOOST_CHECK_EQUAL(UniValue(100000 * COIN).write(), "100000.00000000");
+    BOOST_CHECK_EQUAL(UniValue(10000 * COIN).write(), "10000.00000000");
+    BOOST_CHECK_EQUAL(UniValue(1000 * COIN).write(), "1000.00000000");
+    BOOST_CHECK_EQUAL(UniValue(100 * COIN).write(), "100.00000000");
+    BOOST_CHECK_EQUAL(UniValue(10 * COIN).write(), "10.00000000");
+    BOOST_CHECK_EQUAL(UniValue(COIN).write(), "1.00000000");
+    BOOST_CHECK_EQUAL(UniValue(COIN / 10).write(), "0.10000000");
+    BOOST_CHECK_EQUAL(UniValue(COIN / 100).write(), "0.01000000");
+    BOOST_CHECK_EQUAL(UniValue(COIN / 1000).write(), "0.00100000");
+    BOOST_CHECK_EQUAL(UniValue(COIN / 10000).write(), "0.00010000");
+    BOOST_CHECK_EQUAL(UniValue(COIN / 100000).write(), "0.00001000");
+    BOOST_CHECK_EQUAL(UniValue(COIN / 1000000).write(), "0.00000100");
+    BOOST_CHECK_EQUAL(UniValue(COIN / 10000000).write(), "0.00000010");
+    BOOST_CHECK_EQUAL(UniValue(COIN / 100000000).write(), "0.00000001");
 }
 
 static UniValue ValueFromString(const std::string &str) {
@@ -280,45 +280,47 @@ static UniValue ValueFromString(const std::string &str) {
 }
 
 BOOST_AUTO_TEST_CASE(rpc_parse_monetary_values) {
-    BOOST_CHECK_THROW(AmountFromValue(ValueFromString("-0.01")), UniValue);
+    BOOST_CHECK_THROW(AmountFromValue(ValueFromString("-0.00000001")),
+                      UniValue);
     BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0")), Amount::zero());
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.00")), Amount::zero());
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.01")), SATOSHI);
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("176221.95")),
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.00000000")),
+                      Amount::zero());
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.00000001")), SATOSHI);
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.17622195")),
                       17622195 * SATOSHI);
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("500000")),
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.5")),
                       50000000 * SATOSHI);
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("500000.00")),
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.50000000")),
                       50000000 * SATOSHI);
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("898989.89")),
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.89898989")),
                       89898989 * SATOSHI);
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("1000000.00")),
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("1.00000000")),
                       100000000 * SATOSHI);
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("20999999999999.9")),
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("20999999.9999999")),
                       int64_t(2099999999999990) * SATOSHI);
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("20999999999999.99")),
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("20999999.99999999")),
                       int64_t(2099999999999999) * SATOSHI);
 
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("1e-2")),
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("1e-8")),
                       COIN / 100000000);
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.1e-1")),
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.1e-7")),
                       COIN / 100000000);
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.01e-0")),
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.01e-6")),
                       COIN / 100000000);
     BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString(
                           "0."
                           "0000000000000000000000000000000000000000000000000000"
-                          "000000000000000000000001e+74")),
+                          "000000000000000000000001e+68")),
                       COIN / 100000000);
     BOOST_CHECK_EQUAL(
         AmountFromValue(ValueFromString("10000000000000000000000000000000000000"
-                                        "000000000000000000000000000e-58")),
+                                        "000000000000000000000000000e-64")),
         COIN);
     BOOST_CHECK_EQUAL(
         AmountFromValue(ValueFromString(
             "0."
             "000000000000000000000000000000000000000000000000000000000000000100"
-            "000000000000000000000000000000000000000000000000000e70")),
+            "000000000000000000000000000000000000000000000000000e64")),
         COIN);
 
     // should fail
@@ -327,22 +329,23 @@ BOOST_AUTO_TEST_CASE(rpc_parse_monetary_values) {
     BOOST_CHECK_THROW(AmountFromValue(ValueFromString("0.000000019")),
                       UniValue);
     // should pass, cut trailing 0
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.01000000")), SATOSHI);
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.00000001000000")),
+                      SATOSHI);
     // should fail
     BOOST_CHECK_THROW(AmountFromValue(ValueFromString("19e-9")), UniValue);
     // should pass, leading 0 is present
-    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.19e-0")),
+    BOOST_CHECK_EQUAL(AmountFromValue(ValueFromString("0.19e-6")),
                       19 * SATOSHI);
 
     // overflow error
     BOOST_CHECK_THROW(AmountFromValue(ValueFromString("92233720368.54775808")),
                       UniValue);
     // overflow error
-    BOOST_CHECK_THROW(AmountFromValue(ValueFromString("1e+17")), UniValue);
+    BOOST_CHECK_THROW(AmountFromValue(ValueFromString("1e+11")), UniValue);
     // overflow error signless
-    BOOST_CHECK_THROW(AmountFromValue(ValueFromString("1e17")), UniValue);
+    BOOST_CHECK_THROW(AmountFromValue(ValueFromString("1e11")), UniValue);
     // overflow error
-    BOOST_CHECK_THROW(AmountFromValue(ValueFromString("93e+15")), UniValue);
+    BOOST_CHECK_THROW(AmountFromValue(ValueFromString("93e+9")), UniValue);
 }
 
 BOOST_AUTO_TEST_CASE(json_parse_errors) {
@@ -356,7 +359,7 @@ BOOST_AUTO_TEST_CASE(json_parse_errors) {
     BOOST_CHECK_THROW(AmountFromValue(ParseNonRFCJSONValue(".19e-6")),
                       std::runtime_error);
     BOOST_CHECK_EQUAL(AmountFromValue(ParseNonRFCJSONValue(
-                          "0.00000000000000000000000000000001e+30 ")),
+                          "0.00000000000000000000000000000000000001e+30 ")),
                       SATOSHI);
     // Invalid, initial garbage
     BOOST_CHECK_THROW(ParseNonRFCJSONValue("[1.0"), std::runtime_error);
diff --git a/src/test/util/setup_common.cpp b/src/test/util/setup_common.cpp
index 4ba16508c..c36694eea 100644
--- a/src/test/util/setup_common.cpp
+++ b/src/test/util/setup_common.cpp
@@ -99,7 +99,7 @@ BasicTestingSetup::BasicTestingSetup(
             "-debug",
             "-debugexclude=libevent",
             "-debugexclude=leveldb",
-            "-ecash",
+            "-useecashprefix"
         },
         extra_args);
     // clang-format on
diff --git a/src/test/util_tests.cpp b/src/test/util_tests.cpp
index 4ad8568ed..49cfbb1ab 100644
--- a/src/test/util_tests.cpp
+++ b/src/test/util_tests.cpp
@@ -1359,27 +1359,26 @@ BOOST_AUTO_TEST_CASE(util_ReadWriteSettings) {
 
 BOOST_AUTO_TEST_CASE(util_FormatMoney) {
     BOOST_CHECK_EQUAL(FormatMoney(Amount::zero()), "0.00");
-    BOOST_CHECK_EQUAL(FormatMoney(123456789 * (COIN / 10000)),
-                      "12345678900.00");
-    BOOST_CHECK_EQUAL(FormatMoney(-1 * COIN), "-1000000.00");
-
-    BOOST_CHECK_EQUAL(FormatMoney(100000000 * COIN), "100000000000000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(10000000 * COIN), "10000000000000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(1000000 * COIN), "1000000000000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(100000 * COIN), "100000000000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(10000 * COIN), "10000000000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(1000 * COIN), "1000000000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(100 * COIN), "100000000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(10 * COIN), "10000000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(COIN), "1000000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(COIN / 10), "100000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(COIN / 100), "10000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(COIN / 1000), "1000.00");
-    BOOST_CHECK_EQUAL(FormatMoney(COIN / 10000), "100.00");
-    BOOST_CHECK_EQUAL(FormatMoney(COIN / 100000), "10.00");
-    BOOST_CHECK_EQUAL(FormatMoney(COIN / 1000000), "1.00");
-    BOOST_CHECK_EQUAL(FormatMoney(COIN / 10000000), "0.10");
-    BOOST_CHECK_EQUAL(FormatMoney(COIN / 100000000), "0.01");
+    BOOST_CHECK_EQUAL(FormatMoney(123456789 * (COIN / 10000)), "12345.6789");
+    BOOST_CHECK_EQUAL(FormatMoney(-1 * COIN), "-1.00");
+
+    BOOST_CHECK_EQUAL(FormatMoney(100000000 * COIN), "100000000.00");
+    BOOST_CHECK_EQUAL(FormatMoney(10000000 * COIN), "10000000.00");
+    BOOST_CHECK_EQUAL(FormatMoney(1000000 * COIN), "1000000.00");
+    BOOST_CHECK_EQUAL(FormatMoney(100000 * COIN), "100000.00");
+    BOOST_CHECK_EQUAL(FormatMoney(10000 * COIN), "10000.00");
+    BOOST_CHECK_EQUAL(FormatMoney(1000 * COIN), "1000.00");
+    BOOST_CHECK_EQUAL(FormatMoney(100 * COIN), "100.00");
+    BOOST_CHECK_EQUAL(FormatMoney(10 * COIN), "10.00");
+    BOOST_CHECK_EQUAL(FormatMoney(COIN), "1.00");
+    BOOST_CHECK_EQUAL(FormatMoney(COIN / 10), "0.10");
+    BOOST_CHECK_EQUAL(FormatMoney(COIN / 100), "0.01");
+    BOOST_CHECK_EQUAL(FormatMoney(COIN / 1000), "0.001");
+    BOOST_CHECK_EQUAL(FormatMoney(COIN / 10000), "0.0001");
+    BOOST_CHECK_EQUAL(FormatMoney(COIN / 100000), "0.00001");
+    BOOST_CHECK_EQUAL(FormatMoney(COIN / 1000000), "0.000001");
+    BOOST_CHECK_EQUAL(FormatMoney(COIN / 10000000), "0.0000001");
+    BOOST_CHECK_EQUAL(FormatMoney(COIN / 100000000), "0.00000001");
 }
 
 BOOST_AUTO_TEST_CASE(util_ParseMoney) {
@@ -1387,52 +1386,57 @@ BOOST_AUTO_TEST_CASE(util_ParseMoney) {
     BOOST_CHECK(ParseMoney("0.0", ret));
     BOOST_CHECK_EQUAL(ret, Amount::zero());
 
-    BOOST_CHECK(ParseMoney("1234567.89", ret));
-    BOOST_CHECK_EQUAL(ret, 123456789 * SATOSHI);
+    BOOST_CHECK(ParseMoney("12345.6789", ret));
+    BOOST_CHECK_EQUAL(ret, 123456789 * (COIN / 10000));
 
-    const auto XEC = Currency::get().baseunit;
     BOOST_CHECK(ParseMoney("100000000.00", ret));
-    BOOST_CHECK_EQUAL(ret, 100000000 * XEC);
+    BOOST_CHECK_EQUAL(ret, 100000000 * COIN);
     BOOST_CHECK(ParseMoney("10000000.00", ret));
-    BOOST_CHECK_EQUAL(ret, 10000000 * XEC);
+    BOOST_CHECK_EQUAL(ret, 10000000 * COIN);
     BOOST_CHECK(ParseMoney("1000000.00", ret));
-    BOOST_CHECK_EQUAL(ret, 1000000 * XEC);
+    BOOST_CHECK_EQUAL(ret, 1000000 * COIN);
     BOOST_CHECK(ParseMoney("100000.00", ret));
-    BOOST_CHECK_EQUAL(ret, 100000 * XEC);
+    BOOST_CHECK_EQUAL(ret, 100000 * COIN);
     BOOST_CHECK(ParseMoney("10000.00", ret));
-    BOOST_CHECK_EQUAL(ret, 10000 * XEC);
+    BOOST_CHECK_EQUAL(ret, 10000 * COIN);
     BOOST_CHECK(ParseMoney("1000.00", ret));
-    BOOST_CHECK_EQUAL(ret, 1000 * XEC);
+    BOOST_CHECK_EQUAL(ret, 1000 * COIN);
     BOOST_CHECK(ParseMoney("100.00", ret));
-    BOOST_CHECK_EQUAL(ret, 100 * XEC);
+    BOOST_CHECK_EQUAL(ret, 100 * COIN);
     BOOST_CHECK(ParseMoney("10.00", ret));
-    BOOST_CHECK_EQUAL(ret, 10 * XEC);
+    BOOST_CHECK_EQUAL(ret, 10 * COIN);
     BOOST_CHECK(ParseMoney("1.00", ret));
-    BOOST_CHECK_EQUAL(ret, XEC);
+    BOOST_CHECK_EQUAL(ret, COIN);
     BOOST_CHECK(ParseMoney("1", ret));
-    BOOST_CHECK_EQUAL(ret, XEC);
+    BOOST_CHECK_EQUAL(ret, COIN);
     BOOST_CHECK(ParseMoney("   1", ret));
-    BOOST_CHECK_EQUAL(ret, XEC);
+    BOOST_CHECK_EQUAL(ret, COIN);
     BOOST_CHECK(ParseMoney("1   ", ret));
-    BOOST_CHECK_EQUAL(ret, XEC);
+    BOOST_CHECK_EQUAL(ret, COIN);
     BOOST_CHECK(ParseMoney("  1 ", ret));
-    BOOST_CHECK_EQUAL(ret, XEC);
+    BOOST_CHECK_EQUAL(ret, COIN);
     BOOST_CHECK(ParseMoney("0.1", ret));
-    BOOST_CHECK_EQUAL(ret, XEC / 10);
+    BOOST_CHECK_EQUAL(ret, COIN / 10);
     BOOST_CHECK(ParseMoney("0.01", ret));
-    BOOST_CHECK_EQUAL(ret, XEC / 100);
-    BOOST_CHECK(ParseMoney(" 0.01 ", ret));
-    BOOST_CHECK_EQUAL(ret, XEC / 100);
-    BOOST_CHECK(ParseMoney("0.01 ", ret));
-    BOOST_CHECK_EQUAL(ret, XEC / 100);
-    BOOST_CHECK(ParseMoney(" 0.01", ret));
-    BOOST_CHECK_EQUAL(ret, XEC / 100);
-    BOOST_CHECK(!ParseMoney("0.001", ret));
-    BOOST_CHECK(!ParseMoney("0.0001", ret));
-    BOOST_CHECK(!ParseMoney("0.00001", ret));
-    BOOST_CHECK(!ParseMoney("0.000001", ret));
-    BOOST_CHECK(!ParseMoney("0.0000001", ret));
-    BOOST_CHECK(!ParseMoney("0.00000001", ret));
+    BOOST_CHECK_EQUAL(ret, COIN / 100);
+    BOOST_CHECK(ParseMoney("0.001", ret));
+    BOOST_CHECK_EQUAL(ret, COIN / 1000);
+    BOOST_CHECK(ParseMoney("0.0001", ret));
+    BOOST_CHECK_EQUAL(ret, COIN / 10000);
+    BOOST_CHECK(ParseMoney("0.00001", ret));
+    BOOST_CHECK_EQUAL(ret, COIN / 100000);
+    BOOST_CHECK(ParseMoney("0.000001", ret));
+    BOOST_CHECK_EQUAL(ret, COIN / 1000000);
+    BOOST_CHECK(ParseMoney("0.0000001", ret));
+    BOOST_CHECK_EQUAL(ret, COIN / 10000000);
+    BOOST_CHECK(ParseMoney("0.00000001", ret));
+    BOOST_CHECK_EQUAL(ret, COIN / 100000000);
+    BOOST_CHECK(ParseMoney(" 0.00000001 ", ret));
+    BOOST_CHECK_EQUAL(ret, COIN / 100000000);
+    BOOST_CHECK(ParseMoney("0.00000001 ", ret));
+    BOOST_CHECK_EQUAL(ret, COIN / 100000000);
+    BOOST_CHECK(ParseMoney(" 0.00000001", ret));
+    BOOST_CHECK_EQUAL(ret, COIN / 100000000);
 
     // Parsing amount that can not be represented in ret should fail
     BOOST_CHECK(!ParseMoney("0.000000001", ret));
diff --git a/test/functional/test_framework/test_node.py b/test/functional/test_framework/test_node.py
index ed9aa0c2b..a07e85eab 100755
--- a/test/functional/test_framework/test_node.py
+++ b/test/functional/test_framework/test_node.py
@@ -108,6 +108,7 @@ class TestNode():
             "-debugexclude=leveldb",
             "-uacomment=" + self.name,
             "-noprinttoconsole",
+            "-useecashprefix",
         ]
 
         if use_valgrind:
-- 
2.39.5 (Apple Git-154)

